### PR TITLE
Fix Match.group problems

### LINKED ISSUE #3171: Wrong return type for Match.group()
Return type annotations for methods `Match.group()` should reflect the fact that `None` can be returned, see https://docs.python.org/3/library/re.html#re.Match.group.

https://github.com/python/typeshed/blob/0dea39cb3c07f04abe5b15491e4bd62ce3e6739c/stdlib/3/typing.pyi#L529

### PR DESCRIPTION
**Pull request by srittau**
* group() may return None values
* group() allows to mix int and str arguments
* Use concrete return type
* Merge overloads

Fixes #3171 

Returning `Optional` is potentially problematic. It is often obvious from a regexp whether a certain group exists (every time it is non-optional). Many users may rely on `group()` not returning an `Optional`. Nevertheless, returning `AnyStr` is clearly wrong. One alternative is returning `Any`, but that would lose any type checking ability. Personally, I believe this patch is the best solution, despite the problems it may cause.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Why do you say it is "clearly wrong" to return AnyStr? As far as I can tell, the only problem with that is that we don't handle the `None` case. I can see arguments either way on that one, but it's probably best to err on the side of type safety here and include `None` in the return type.

Could you also make the arguments positional-only?

```
  File "<ipython-input-44-e82d21244346>", line 1, in <module>
    m.group(group1='1')
TypeError: group() takes no keyword arguments
```

**Comment by srittau** [comment]
Well, "clearly wrong" -> "not type-safe in the general case". Will change them to positional only.

File: stdlib/3/typing.pyi [hunk 1]
```diff
@@ -525,10 +525,16 @@ class Match(Generic[AnyStr]):
 
     def expand(self, template: AnyStr) -> AnyStr: ...
 
+    def group(self, __group: Literal[0] = ...) -> AnyStr: ...
```
Comment (JelleZijlstra) [inline stdlib/3/typing.pyi]: Missing the `@overload` decorator

**Comment by ilevkivskyi** [comment]
This causes 19 errors in mypy self-check and 45 errors in the Python 3 part of our internal code base (which is actually relatively small).

TBH, I am not sure the `Optional` part in the second overload is really worth it. What do you think about (partially) reverting this?

**Comment by srittau** [comment]
I wonder why our tests did not catch this, since the mypy self-test is run as part of the build. I'm open to changing it back, although just removing `Optional` can problems in theory as well. For example, a code analysis tool making use of type annotations could complain about the "impossible" branch in code like this:

```python
if m.group(12) is None:
    ...
```

While I don't think such a tool exists yet, I'd prefer to use `Any` if we decide to change it.

**Comment by ilevkivskyi** [comment]
Actually I was wrong about number of errors internally, it is not 45 it is 110. So I think it makes sense reverting this in some form.

As for the `--warn-unreachable` problem, this problem indeed exists, but as practice shows, it is still quite rare with the old signature. For example it caused only couple errors in mypy itself. So I think reverting to `AnyStr` is a safe bet.

**Pull request merged**